Functions/scope_lexical_scope_and_closures.md

```diff
@@ -0,0 +1,107 @@
+-   Variables declared with var, let and const are quite similar when declared inside a function.
+
+#### Function Scope vs. Block Scope:
+
+-   Introduction of let and const in ES6 brought block-level scoping.
+-   let and const variables are block-scoped, unlike var which is function-scoped.
+
+#### Scope Chain
+
+Scope also follows a hierarchy known as the scope chain. When a variable is referenced in JavaScript, the JavaScript engine looks for that variable first in the current scope. If it's not found, it traverses up the scope chain until it finds the variable or reaches the global scope. This process ensures that variables are accessed in the correct scope.
+
+Understanding scope is essential for writing maintainable and predictable JavaScript code. It helps prevent naming conflicts, promotes code organization, and improves code readability. Additionally, understanding scope is crucial for understanding closures, hoisting, and other advanced JavaScript concepts.
+
+## Lexical context
+
+**Lexical context**, also known as lexical scope, refers to the environment in which a piece of code is written or defined. In JavaScript, lexical scope is determined by the physical placement of code within the source code file. This means that variables and functions declared within a certain scope are accessible within that scope and any nested scopes.
+
+When we talk about lexically binding the **this** value in JavaScript, it means that the value of this inside a function is determined by the function's lexical scope, rather than how the function is called. This behavior is particularly relevant when dealing with nested functions or functions used as callbacks, where the value of this might change depending on how the function is invoked.
+
+In traditional functions (non-arrow functions), the value of this is dynamically determined at runtime based on how the function is called. This can lead to unexpected behavior, especially in nested functions or when passing functions as callbacks.
+
+However, arrow functions in JavaScript lexically bind the this value, meaning that this inside an arrow function is determined by the surrounding lexical scope where the arrow function is defined. This behavior ensures that the this value inside an arrow function is always the same as the this value in the surrounding code.
+
+Consider the following example to illustrate the difference:
+
+```js
+const obj = {
+    name: "Alice",
+    traditionalFunc: function () {
+        console.log(this.name); // 'Alice'
+        setTimeout(function () {
+            console.log(this.name); // undefined (or error in strict mode)
+        }, 1000);
+    },
+    arrowFunc: function () {
+        console.log(this.name); // 'Alice'
+        setTimeout(() => {
+            console.log(this.name); // 'Alice'
+        }, 1000);
+    },
+};
+
+obj.traditionalFunc(); // Executes with traditional function
+obj.arrowFunc(); // Executes with arrow function
+```
+
+## Closures
```

Comment:
could u write some example of a closure please?

Reply:
Sure!